core/os/android/adb/perfetto.go

```diff
@@ -29,6 +29,95 @@ import (
 	"github.com/google/gapid/core/log"
 )
 
+const (
+	perfettoProducerLaunchersDirectory = "/data/local/tmp/perfetto_producer_launchers"
+)
+
+func (b *binding) preparePerfettoProducerLauncherFromApk(ctx context.Context, packageName string, launcherBinary string) (string, error) {
+	if b.Instance().GetConfiguration().GetOS().GetAPIVersion() < 29 {
+		return "", log.Errf(ctx, nil, "Querying perfetto capability requires Android API >= 29")
+	}
+	launcherPath := perfettoProducerLaunchersDirectory + "/" + launcherBinary
+	if _, err := b.Shell("rm", "-f", launcherPath).Call(ctx); err != nil {
+		return "", log.Errf(ctx, err, "Can't clean up existing perfetto producer launcher %v", launcherBinary)
+	}
+
+	// Attempt to create the directory
+	b.Shell("mkdir", "-p", perfettoProducerLaunchersDirectory).Call(ctx)
+	res, err := b.Shell("pm", "path", packageName).Call(ctx)
+	if err != nil {
+		return "", log.Errf(ctx, err, "Failed to query path to apk %v", packageName)
+	}
+	packagePath := strings.Split(res, ":")[1]
+	if _, err := b.Shell("unzip", "-o", packagePath, "assets/"+launcherBinary, "-d", perfettoProducerLaunchersDirectory).Call(ctx); err != nil {
+		return "", log.Errf(ctx, err, "Failed to unzip %v from %v", launcherBinary, packageName)
+	}
+
+	// unzip also creates the directory structure, clean it up.
+	b.Shell("mv", perfettoProducerLaunchersDirectory+"/assets/"+launcherBinary, perfettoProducerLaunchersDirectory).Call(ctx)
+	b.Shell("rm", "-rf", perfettoProducerLaunchersDirectory+"/assets").Call(ctx)
+
+	// Finally, make sure the binary is executable
+	b.Shell("chmod", "a+x", launcherPath).Call(ctx)
+	return launcherPath, nil
+}
+
+func (b *binding) LaunchPerfettoProducerFromApk(ctx context.Context, packageName string, launcherBinary string, started chan int) error {
+	// Firstly, extract the producer launcher from Apk.
+	binaryPath, err := b.preparePerfettoProducerLauncherFromApk(ctx, packageName, launcherBinary)
+	if err != nil {
+		return err
+	}
+
+	// Construct IO pipe, shell command outputs to stdout, GAPID reads from
+	// reader for logging purpose.
+	reader, stdout := io.Pipe()
+	fail := make(chan error, 1)
+	crash.Go(func() {
+		buf := bufio.NewReader(reader)
+		for {
+			line, e := buf.ReadString('\n')
+			// As long as there's output, consider the binary starting running.
+			started <- 1
```

Comment:
after the second line this will block, since you created the `started` chan with size 1 and only read from it once.

Use our task.Signal instead.